java.net.URISyntaxException: Expected scheme-specific part at index

Looks like we're not catching a possible exception thrown by `URI`:

https://github.com/mozilla-mobile/android-components/blob/14fdb2d78a009bc0dc782cb28898a6f5d3128c23/components/browser/engine-system/src/main/java/mozilla/components/browser/engine/system/SystemEngineView.kt#L131

Some thoughts:
- We're trying to use `android.net.Uri` everywhere based on previous discussions: https://github.com/mozilla-mobile/android-components/issues/648 - maybe this should be Uri?
- ^ I wonder if we should have static analysis to ensure we're only using `Uri` and not using `URI` or `URL` or whatever else
- Reminder to add a regression test! :)

This is an issue in firefox-tv v3.0.

```
     Caused by: java.net.URISyntaxException: Expected scheme-specific part at index 4: fgh:
        at java.net.URI$Parser.fail(URI.java:2855)
        at java.net.URI$Parser.failExpecting(URI.java:2861)
        at java.net.URI$Parser.parse(URI.java:3064)
        at java.net.URI.<init>(URI.java:589)
        at mozilla.components.browser.engine.system.SystemEngineView$createWebViewClient$1$onPageFinished$$inlined$let$lambda$1.invoke(SystemEngineView.kt:130)
        at mozilla.components.browser.engine.system.SystemEngineView$createWebViewClient$1$onPageFinished$$inlined$let$lambda$1.invoke(SystemEngineView.kt:113)
        at mozilla.components.support.base.observer.ObserverRegistry.notifyObservers(ObserverRegistry.kt:106)
        at mozilla.components.concept.engine.EngineSession.notifyObservers(EngineSession.kt)
        at mozilla.components.browser.engine.system.SystemEngineSession.internalNotifyObservers$engine_system_release(SystemEngineSession.kt:278)
        at mozilla.components.browser.engine.system.SystemEngineView$createWebViewClient$1.onPageFinished(SystemEngineView.kt:127)
        at com.amazon.webview.awvclients.AmazonWebViewClient.onPageFinished(AmazonWebViewClient.java:95)
        at com.android.webview.chromium.WebViewContentsClientAdapter.onPageFinished(WebViewContentsClientAdapter.java:546)
        at com.android.webview.chromium.WebViewExtAmazonContentsClientAdapter.onPageFinished(WebViewExtAmazonContentsClientAdapter.java:18)
        at org.chromium.android_webview.AwContentsClientCallbackHelper$MyHandler.handleMessage(AwContentsClientCallbackHelper.java:201)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:154)
```

--- a/components/browser/engine-system/src/main/java/mozilla/components/browser/engine/system/SystemEngineView.kt
+++ b/components/browser/engine-system/src/main/java/mozilla/components/browser/engine/system/SystemEngineView.kt
@@ -38,7 +38,6 @@ import mozilla.components.concept.engine.HitResult
 import mozilla.components.support.ktx.android.content.isOSOnLowMemory
 import mozilla.components.support.utils.DownloadUtils
 import java.lang.ref.WeakReference
-import java.net.URI
 
 /**
  * WebView-based implementation of EngineView.
@@ -128,7 +127,10 @@ class SystemEngineView @JvmOverloads constructor(
                 session?.internalNotifyObservers {
                     onLocationChange(it)
                     onLoadingStateChange(false)
-                    onSecurityChange(cert != null, cert?.let { URI(url).host }, cert?.issuedBy?.oName)
+                    onSecurityChange(
+                            secure = cert != null,
+                            host = cert?.let { Uri.parse(url).host },
+                            issuer = cert?.issuedBy?.oName)
 
                     if (!isLowOnMemory()) {
                         val thumbnail = session?.captureThumbnail()

--- a/components/browser/engine-system/src/test/java/mozilla/components/browser/engine/system/SystemEngineViewTest.kt
+++ b/components/browser/engine-system/src/test/java/mozilla/components/browser/engine/system/SystemEngineViewTest.kt
@@ -77,7 +77,7 @@ class SystemEngineViewTest {
         observedLoadingState = true
         engineView.currentWebView.webViewClient.onPageFinished(null, "http://mozilla.org")
         assertEquals("http://mozilla.org", observedUrl)
-        assertEquals(false, observedLoadingState)
+        assertFalse(observedLoadingState)
         assertEquals(Triple(false, null, null), observedSecurityChange)
 
         val view = mock(WebView::class.java)
@@ -86,6 +86,10 @@ class SystemEngineViewTest {
 
         val certificate = mock(SslCertificate::class.java)
         val dName = mock(SslCertificate.DName::class.java)
+        doReturn("testCA").`when`(dName).oName
+        doReturn(certificate).`when`(view).certificate
+        engineView.currentWebView.webViewClient.onPageFinished(view, "http://mozilla.org")
+
         doReturn("testCA").`when`(dName).oName
         doReturn(dName).`when`(certificate).issuedBy
         doReturn(certificate).`when`(view).certificate
@@ -579,5 +583,37 @@ class SystemEngineViewTest {
         })
         engineView.currentWebView.webViewClient.onPageFinished(null, "http://mozilla.org")
         assertFalse(thumbnailChanged)
+    }
+
+    @Test
+    fun `onPageFinished handles invalid URL`() {
+        val engineSession = SystemEngineSession()
+        val engineView = SystemEngineView(RuntimeEnvironment.application)
+        engineView.render(engineSession)
+
+        var observedUrl = ""
+        var observedLoadingState = true
+        var observedSecurityChange: Triple<Boolean, String?, String?> = Triple(false, null, null)
+        engineSession.register(object : EngineSession.Observer {
+            override fun onLoadingStateChange(loading: Boolean) { observedLoadingState = loading }
+            override fun onLocationChange(url: String) { observedUrl = url }
+            override fun onSecurityChange(secure: Boolean, host: String?, issuer: String?) {
+                observedSecurityChange = Triple(secure, host, issuer)
+            }
+        })
+
+        // We need a certificate to trigger parsing the potentially invalid URL for
+        // the host parameter in onSecurityChange
+        val view = mock(WebView::class.java)
+        val certificate = mock(SslCertificate::class.java)
+        val dName = mock(SslCertificate.DName::class.java)
+        doReturn("testCA").`when`(dName).oName
+        doReturn(dName).`when`(certificate).issuedBy
+        doReturn(certificate).`when`(view).certificate
+
+        engineView.currentWebView.webViewClient.onPageFinished(view, "invalid:")
+        assertEquals("invalid:", observedUrl)
+        assertFalse(observedLoadingState)
+        assertEquals(Triple(true, null, "testCA"), observedSecurityChange)
     }
 }

